cpp/src/arrow/result.h

```diff
@@ -294,7 +294,13 @@ class [[nodiscard]] Result : public util::EqualityComparable<Result<T>> {
   ///
   /// \return The stored non-OK status object, or an OK status if this object
   ///         has a value.
-  constexpr const Status& status() const { return status_; }
+  constexpr const Status& status() const& { return status_; }
+
+  /// Gets the stored status object, or an OK status if a `T` value is stored.
+  ///
+  /// \return The stored non-OK status object, or an OK status if this object
+  ///         has a value.
+  Status status() && { return status_; }
```

Comment:
```suggestion
  Status status() && { return std::move(status_); }
```

Without the move in here, we wind up copying status out of this and into the return value. See also `absl::StatusOr<T>::status()&&`:

https://github.com/abseil/abseil-cpp/blob/master/absl/status/statusor.h#L703

Reply:
(If status is an error, then there is no stored object to destroy. If status is not an error, then it is cheap to construct an ok status to return here)

Reply:
The issue with this is that std::move(status_) will leave status_ in an OK state (since the internal state_ pointer will be NULL afterward). As a result, the Result destructor will attempt to call a destructor on the storage_ field, leading to undefined behavior because storage_ is uninitialized when the original Result was constructed in an error state. One possible solution (assuming we need to minimize the size of Result without simply putting status_ and storage_ in a variant) is for the move constructor of Status to preserve the binary OK/error state of the moved-out object. This could be achieved without changing the size of Status by using pointer tagging for its state_ field. There is a [proposal](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3125r1.html) for this, which will likely make it into C++26, but until then, there’s no truly safe way to do this while complying with the current C++ standard…

Reply:
@bkietz I'd love to hear your feedback on that

Reply:
> The issue with this is that std::move(status_) will leave status_ in an OK state

Please mark the result as an error with a placeholder status and move the error status out. This won't be any more expensive than what you've already written, and we can replace the placeholder with a static status after #44493 
```suggestion
  Status status() && {
    if (ok()) return Status::OK();
    auto out = std::move(status_);
    status_ = Status::UnknownError("Uninitialized Result<T>");
    return out;
  }
```

> There is a [proposal](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3125r1.html) for this, which will likely make it into C++26

Whenever this is ready, it'll make constructs like static error statuses much nicer.

Reply:
Thank you @bkietz. I applied the logic you suggested. Can we mark it resolved?